#VRML_SIM R2023b utf8

PROTO Vehicle [
]

{

DEF ROBOT Robot {

    translation  0 0 0.30 # 0.015
        rotation 0 0 0 0
        controller "<extern>"
        supervisor FALSE
        name "Vehicle"
        model "Custom MAV"

        children [

        DEF BODY Pose {
            translation 0 0 0
                children [

                CadShape {
                    # url [ "../meshes/collada_files/cf2_assembly.dae" ]
                    url [ "../meshes/collada_files/custom.dae" ]
                }

                ]
      }

    Camera {
        translation 0.03 0 0.01
            fieldOfView 0.87
            width 324
            height 324
    }

    GPS {
    }

    Gyro {
    }

    InertialUnit {
        name "inertial_unit"
    }

    # Motor 1
    Propeller {
        shaftAxis 0 0 1
            centerOfThrust -0.031 -0.031 0.008
            thrustConstants -4e-05 0
            torqueConstants 2.4e-06 0
            fastHelixThreshold 50
            device RotationalMotor {
                name "motor1"
                    maxVelocity 100
                    maxTorque 30
            }
        fastHelix Solid {

            translation -0.031 -0.031 0.008

                children [ 

                    Shape {
                            appearance PBRAppearance {
                                baseColor 0.4 0.0 0.0
                                    baseColorMap ImageTexture {
                                        url [
                                            "../meshes/textures/fast_helix.png"
                                        ]
                                    }
                                metalness 0
                            }
                            geometry Cylinder {
                                height 0.001
                                    radius 0.025
                            }
                        }

                ]
        }
        slowHelix Solid {
            translation -0.031 -0.031 0.008
                children [
                    Shape {
                        appearance DEF PLASTIC PBRAppearance {
                            baseColor 1 0 0
                                metalness 0.3
                        }
                        geometry Mesh {
                            url [
                                "../meshes/collada_files/cw_prop.dae"
                            ]
                        }
                    }

                ]
        }
    }

    # Motor 2
    Propeller {
        shaftAxis 0 0 1
            centerOfThrust 0.031 -0.031 0.008  
            thrustConstants 4e-05 0               
            torqueConstants 2.4e-06 0            
            fastHelixThreshold 50                 
            device RotationalMotor {
                name "motor2"
                    maxVelocity 100                     
                    maxTorque 30                        
            }
        fastHelix Solid {
            translation 0.031 -0.031 0.008
                children [ 

                Shape {
                            appearance PBRAppearance {
                                baseColor 0.1 0.1 0.1
                                    baseColorMap ImageTexture {
                                        url [
                                            "../meshes/textures/fast_helix.png"
                                        ]
                                    }
                                metalness 0
                            }
                            geometry Cylinder {
                                height 0.001
                                    radius 0.025
                            }
                        }

                ]
        }
        slowHelix Solid {
            translation 0.031 -0.031 0.008
                children [
                    Shape {
                        appearance DEF PLASTIC PBRAppearance {
                            baseColor 0 0 0
                                metalness 0.3
                        }
                        geometry Mesh {
                            url [
                                "../meshes/collada_files/ccw_prop.dae"
                            ]
                        }
                    }
                ]
        }
    }

    # Motor 3
    Propeller {
        shaftAxis 0 0 1
            centerOfThrust -0.031 0.031 0.008
            thrustConstants 4e-05 0
            torqueConstants 2.4e-06 0
            fastHelixThreshold 50
            device RotationalMotor {
                name "motor3"
                    maxVelocity 100
                    maxTorque 30
            }
        fastHelix Solid {
            translation -0.031 0.031 0.008
                children [ 
                  
                Shape {
                            appearance PBRAppearance {
                                baseColor 0.4 0.0 0.0
                                    baseColorMap ImageTexture {
                                        url [
                                            "../meshes/textures/fast_helix.png"
                                        ]
                                    }
                                metalness 0
                            }
                            geometry Cylinder {
                                height 0.001
                                    radius 0.025
                            }
                        }

                ]
        }
        slowHelix Solid {
            translation -0.031 0.031 0.008
                children [
                    Shape {
                        appearance DEF PLASTIC PBRAppearance {
                            baseColor 1 0 0
                                metalness 0.3
                        }
                        geometry Mesh {
                            url [
                                "../meshes/collada_files/ccw_prop.dae"
                            ]
                        }
                    }

                ]
        }
    }

    # Motor 4
    Propeller {
        shaftAxis 0 0 1
            centerOfThrust 0.031 0.031 0.008
            thrustConstants -4e-05 0
            torqueConstants 2.4e-06 0
            fastHelixThreshold 50
            device RotationalMotor {
                name "motor4"
                    maxVelocity 100  # rad/sec
                    maxTorque 30
            }
        fastHelix Solid {
            translation 0.031 0.031 0.008
                children [

                Shape {
                            appearance PBRAppearance {
                                baseColor 0.1 0.1 0.1
                                    baseColorMap ImageTexture {
                                        url [
                                            "../meshes/textures/fast_helix.png"
                                        ]
                                    }
                                metalness 0
                            }
                            geometry Cylinder {
                                height 0.001
                                    radius 0.025
                            }
                        }

              
                ] 
        }
        slowHelix DEF PROP Solid {
            translation 0.031 0.031 0.007
                children [
                    Shape {
                        appearance DEF PLASTIC PBRAppearance {
                            baseColor 0 0 0
                                metalness 0.3
                        }
                        geometry Mesh {
                            url [
                                "../meshes/collada_files/cw_prop.dae"
                            ]
                        }
                    }

                ]
        }
    }
    ]
        boundingObject Cylinder {
            height 0.03
                radius 0.05
        }
    physics Physics {
        density -1
            mass 0.05
    }
}

}
